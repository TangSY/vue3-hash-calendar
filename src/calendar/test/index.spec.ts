import { Calendar } from '..';
import { mount, later } from '../../../test';
import {
  defaultMultipleDate,
  defaultRangeDate,
  defaultYearMonth,
  ON_CHANGE_BASIC_CALLED_TIMES,
} from './utils';

test('range-type with allow-same-day', async () => {
  const onChange = jest.fn();
  const wrapper = mount(Calendar, {
    props: {
      selectType: 'range',
      defaultYearMonth,
      onChange,
    },
  });
  await later(200);
  expect(onChange).toHaveBeenCalledTimes(ON_CHANGE_BASIC_CALLED_TIMES);

  const days = wrapper.findAll('.calendar_item');
  await days[55].trigger('click');
  expect(onChange).toHaveBeenCalledTimes(ON_CHANGE_BASIC_CALLED_TIMES);
  await days[55].trigger('click');
  expect(onChange).toHaveBeenCalledTimes(ON_CHANGE_BASIC_CALLED_TIMES);

  wrapper.setProps({ allowSameDay: true });
  await days[55].trigger('click');
  expect(onChange).toHaveBeenCalledTimes(ON_CHANGE_BASIC_CALLED_TIMES);
  await days[55].trigger('click');
  expect(onChange).toHaveBeenCalledTimes(ON_CHANGE_BASIC_CALLED_TIMES + 1);
  expect(onChange).toHaveBeenLastCalledWith([
    new Date(2022, 0, 1),
    new Date(2022, 0, 1),
  ]);
  expect(days[55].attributes('class')).toContain('calendar_range_start-end');
});

test('range-type with defaultDatetime', async () => {
  const onChange = jest.fn();
  const wrapper = mount(Calendar, {
    props: {
      selectType: 'range',
      defaultDatetime: defaultRangeDate,
      defaultYearMonth,
      onChange,
    },
  });
  await later(200);

  const days = wrapper.findAll('.calendar_item');
  expect(days[55].attributes('class')).toContain('calendar_range_start');
  expect(days[56].attributes('class')).toContain('calendar_range_middle');
  expect(days[57].attributes('class')).toContain('calendar_range_middle');
  expect(days[58].attributes('class')).toContain('calendar_range_middle');
  expect(days[59].attributes('class')).toContain('calendar_range_end');
});

test('range-type with click', async () => {
  const onChange = jest.fn();
  const onClick = jest.fn();
  const wrapper = mount(Calendar, {
    props: {
      selectType: 'range',
      defaultYearMonth,
      onChange,
      onClick,
    },
  });
  await later(200);
  expect(onChange).toHaveBeenCalledTimes(ON_CHANGE_BASIC_CALLED_TIMES);

  const days = wrapper.findAll('.calendar_item');
  await days[55].trigger('click');
  expect(onClick).toHaveBeenLastCalledWith(new Date(2022, 0, 1));
  expect(onChange).toHaveBeenCalledTimes(ON_CHANGE_BASIC_CALLED_TIMES);
  await days[57].trigger('click');
  expect(onClick).toHaveBeenLastCalledWith(new Date(2022, 0, 3));
  expect(onChange).toHaveBeenCalledTimes(ON_CHANGE_BASIC_CALLED_TIMES + 1);
  expect(onChange).toHaveBeenLastCalledWith([
    new Date(2022, 0, 1),
    new Date(2022, 0, 2),
    new Date(2022, 0, 3),
  ]);

  await days[58].trigger('click');
  expect(onClick).toHaveBeenLastCalledWith(new Date(2022, 0, 4));
  expect(onChange).toHaveBeenCalledTimes(ON_CHANGE_BASIC_CALLED_TIMES + 1);
  expect(wrapper.find('.calendar_range_start').exists()).toBeTruthy();
  expect(wrapper.find('.calendar_range_middle').exists()).toBeFalsy();
  expect(wrapper.find('.calendar_range_end').exists()).toBeFalsy();

  await days[56].trigger('click');
  expect(onChange).toHaveBeenCalledTimes(ON_CHANGE_BASIC_CALLED_TIMES + 1);
  await days[59].trigger('click');
  expect(onChange).toHaveBeenCalledTimes(ON_CHANGE_BASIC_CALLED_TIMES + 2);
  expect(onChange).toHaveBeenLastCalledWith([
    new Date(2022, 0, 2),
    new Date(2022, 0, 3),
    new Date(2022, 0, 4),
    new Date(2022, 0, 5),
  ]);
});

test('multiple-type with defaultDatetime', async () => {
  const onChange = jest.fn();
  const wrapper = mount(Calendar, {
    props: {
      selectType: 'multiple',
      defaultDatetime: defaultMultipleDate,
      defaultYearMonth,
      onChange,
    },
  });
  await later(200);

  const days = wrapper.findAll('.calendar_day');
  expect(days[55].attributes('class')).toContain('calendar_day_checked');
  expect(days[56].attributes('class')).not.toContain('calendar_day_checked');
  expect(days[59].attributes('class')).toContain('calendar_day_checked');
  expect(days[60].attributes('class')).not.toContain('calendar_day_checked');
  expect(days[64].attributes('class')).toContain('calendar_day_checked');
});

test('multiple-type with click', async () => {
  const onChange = jest.fn();
  const onClick = jest.fn();
  const wrapper = mount(Calendar, {
    props: {
      selectType: 'multiple',
      defaultDatetime: defaultMultipleDate,
      defaultYearMonth,
      onChange,
      onClick,
    },
  });
  await later(200);

  const days = wrapper.findAll('.calendar_day');
  expect(onChange).toHaveBeenCalledTimes(ON_CHANGE_BASIC_CALLED_TIMES);
  await days[55].trigger('click');
  expect(onClick).toHaveBeenLastCalledWith(new Date(2022, 0, 1));
  expect(onChange).toHaveBeenLastCalledWith([
    new Date(2022, 0, 5),
    new Date(2022, 0, 10),
  ]);
  expect(onChange).toHaveBeenCalledTimes(ON_CHANGE_BASIC_CALLED_TIMES + 1);
  await days[59].trigger('click');
  expect(onChange).toHaveBeenLastCalledWith([new Date(2022, 0, 10)]);
  await days[64].trigger('click');
  expect(onChange).toHaveBeenLastCalledWith([]);
  await days[55].trigger('click');
  expect(onChange).toHaveBeenLastCalledWith([new Date(2022, 0, 1)]);
});
