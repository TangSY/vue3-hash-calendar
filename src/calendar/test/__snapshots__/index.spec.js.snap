// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render demo button 1`] = `
<button class="demo-button">
</button>
`;
